File: typescript/packages/var-core/src/step-role.ts
// The role a step definition plays, mirroring concepts/sensors-and-actuators.md:
//   stimulus — drives the software: arranges the quiescent state AND acts on it
//   sensor   — the read-only assertion (the only role that returns for comparison)
// The concepts arrange/act (given/when) remain useful narration in a document,
// but they share one mechanism: a stimulus evolves state, a sensor observes it.
export type StepKind = 'stimulus' | 'sensor'

// Guess a step's role from its neighbours, using the canonical document order
// stimulus → sensor. Purely structural — never inspects sentence words (no
// Given/When/Then heuristics). The generated snippet always offers the other
// role as a commented alternative, so a wrong guess is cheap to correct.
export function inferStepRole(neighbours: {
  readonly before: ReadonlyArray<StepKind>
  readonly after: ReadonlyArray<StepKind>
}): StepKind {
  // A step with nothing after it is most likely the observation; anything
  // followed by other steps is most likely driving the software.
  return neighbours.after.length === 0 ? 'sensor' : 'stimulus'
}

File: java/var-core/src/main/java/dev/varar/core/StepRole.java
package com.oselvar.var.core;

import java.util.List;

/**
 * Port of {@code var-core/src/step-role.ts}: guess a step's role from its neighbours in
 * document order.
 */
public final class StepRole {

    private StepRole() {}

    /**
     * The kinds of the steps immediately {@code before} and {@code after} the step being
     * inferred, in document order.
     */
    public record Neighbours(List<StepKind> before, List<StepKind> after) {
        public Neighbours {
            before = List.copyOf(before);
            after = List.copyOf(after);
        }
    }

    /**
     * Guesses a step's role from its neighbours, using the canonical document order
     * stimulus &rarr; sensor. Purely structural — never inspects sentence words (no
     * Given/When/Then heuristics). The generated snippet always offers the other role
     * as a commented alternative, so a wrong guess is cheap to correct.
     *
     * <p>A step with nothing after it is most likely the observation; anything followed
     * by other steps is most likely driving the software.
     */
    public static StepKind inferStepRole(Neighbours neighbours) {
        return neighbours.after().isEmpty() ? StepKind.SENSOR : StepKind.STIMULUS;
    }
}
